pkg/test/echo/server/forwarder/http.go

```diff
@@ -119,7 +120,12 @@ func newHTTP2TransportGetter(cfg *Config) (httpTransportGetter, func()) {
 			// So http2.Transport doesn't complain the URL scheme isn't 'https'
 			AllowHTTP: true,
 			// Pretend we are dialing a TLS endpoint. (Note, we ignore the passed tls.Config)
-			DialTLS: func(network, addr string, _ *tls.Config) (net.Conn, error) {
+			DialTLSContext: func(ctx context.Context, network, addr string, _ *tls.Config) (net.Conn, error) {
+				if strings.Contains(cfg.Proxy, "socks5") {
+					splitProxyURL := strings.Split(cfg.Proxy, "://")
+					dialer, _ := proxy.SOCKS5("tcp", splitProxyURL[1], nil, proxy.Direct)
+					return dialer.Dial(network, addr)
```

Comment:
I am not sure this is the appropriate place. Why are we using socks5 only for http2?

It seems like it should be in the `newDialer` function

Reply:
hbone.Dialer is just proxy.Dialer + proxy.DialerContext. socks5 actually meets this, so I am pretty sure you could just type cast it.

Its still not clear to me why we use socks5 only for h2c though?

Reply:
@howardjohn So, to make sure I understand your proposal, your argument is that instead of having a separate function newProxyDialer which is called in only the h2 context we should instead put that code into newDialer, so it would look something like this:

```
func newDialer(cfg *Config) hbone.Dialer {

       // This is new
	if proxyURL.Scheme == "socks5" {
		out, _ := proxy.SOCKS5("tcp", proxyURL.Host, nil, proxy.Direct)
		return out
	}
	// End new code
	
	if cfg.Request.Hbone.GetAddress() != "" {
		out := hbone.NewDialer(hbone.Config{
			ProxyAddress: cfg.Request.Hbone.GetAddress(),
			Headers:      cfg.hboneHeaders,
			TLS:          cfg.hboneTLSConfig,
		})
		return out
	}
	out := &net.Dialer{
		Timeout: common.ConnectionTimeout,
	}
	if cfg.forceDNSLookup {
		out.Resolver = newResolver(common.ConnectionTimeout, "", "")
	}
	return out
}
```

Which should allow socks5 to be used as dialer for anything that calls newDialer (http1, http2, but not http3). And this works because proxy.Dialer can be cast to hbone.Dialer without issue.

Is that correct?

Reply:
Yeah